zmq_addon.hpp

```diff
@@ -572,6 +575,62 @@ class active_poller_t
 #endif //  defined(ZMQ_BUILD_DRAFT_API) && defined(ZMQ_CPP11) && defined(ZMQ_HAVE_POLLER)
 
 
+#ifdef ZMQ_CPP11
+/*  A context shutdown scope guard.
+
+    Starts a context shutdown loop in another thread
+    until destructed or explicitly stopped.
+    Useful for unblocking blocking function calls
+    and stopping further socket operations
+    in multi-threaded applications before terminating
+    the context. Example usage:
+
+    {
+        zmq::shutdown_guard sg{ctx};
+        threads.join(); // join all threads using sockets
+    } // stops sg
+    ctx.close();        // terminate context
+*/
+class shutdown_guard
+{
+public:
+    // Call ctx.shutdown() repeatedly on iv intervals until stopped.
+    explicit shutdown_guard(zmq::context_t& ctx,
+                            std::chrono::microseconds iv = std::chrono::microseconds(5000))
+    {
+        thrd = std::thread([&, iv]{
+            while (true)
+            {
+                ctx.shutdown();
+                std::unique_lock<std::mutex> lock{mtx};
+                if (cv.wait_for(lock, iv, [this]{ return do_stop; }))
+                    return; // stop requested
+            }
```

Comment:
I somehow don't get the point on the use case for this. Why does this need to be a loop which might call shutdown multiple times? When would a single call to shutdown not be sufficient?

Reply:
This might not be required if 1 socket is created before shutdown, see https://github.com/zeromq/libzmq/issues/3792. A better workaround/hack is maybe to create a temporary socket in context_t::shutdown before calling zmq_ctx_shutdown?